### PR TITLE
Fix fake toas

### PR DESCRIPTION
**Pull request by dlakaplan**
Making sure that `zima` works correctly, using the same `make_fake_toas` functions as other parts of the code.

This led to some changes:
* Moved the functions (and `calculate_random_models`) to a new module `pint.simulation`.  The reason for this was that including them in `toa` led to circular import worries, and the same for including them in `utils`.  Those imports should really be sorted elsewhere but I didn't want to get into that.
* Added a base function `make_fake_toas` that accepts a `TOAs` object as input.  This then has specific uses:
* `make_fake_toas_uniform` (the old function - uniformly spaced)
* `make_fake_toas_fromtim` (`zima` did this too)
* added the ability to "fuzz" the uniform MJDs (replicating `zima`) 
* added the ability to actually include the error as noise on the TOAs, not just populate a column (again, `zima` did this, but it's now optional for both)
* checked that this works with `EFAC` and `EQUAD`
* added unit tests

I updated all of the existing tests and docs to reflect these changes.  

Usages:
```
toas = pint.simulation.make_fake_toas_uniform(57001, 58000, 200, model=model, error=1 * u.us,add_noise=True )
```
or
```
m, t = get_model_and_toas(
        pint.config.examplefile("B1855+09_NANOGrav_9yv1.gls.par"),
        pint.config.examplefile("B1855+09_NANOGrav_9yv1.tim"),)
# refit the data to get rid of a trend
f = GLSFitter(t, m)
f.fit_toas()
t_sim = pint.simulation.make_fake_toas(t, f.model, add_noise=True)
```

### CODE REVIEW

**Comment by aarchiba** [comment]
How about the basic function accepts a TOAs object as input? And twiddles them minimally to produce zero residuals? And optionally adds noise? I'm thinking of cases that are awkward to generate now like:

- Generate fake TOAs with two different telescopes, with different frequencies, uncertainties, and maybe observing cadences, that fit to the model
- Generate TOAs according to some observing campaign but save a lot of time by not worrying about the residuals (this comes up in testing!)
- Generate narrowband-style TOAs in clumps

Some of this can be helped by `merge_toas` but absolute phase can be a headache. So if `pint.simulate` contained a bunch of functions to produce TOAs simulating various observing campaigns, one could use `merge_toas` to get even more complex campaigns, and then feed the giant mess to the `make_residuals_zero` function. The `add_noise` function here would presumably add the claimed uncertainties; as we were talking about in #1095 we'd eventually add `model.generate_noise(toas)` to handle EQUAD/ECORR/red noise.

**Comment by dlakaplan** [comment]
Having the base function operate on a TOA object (rather than an array of MJDs) could work.  Agree that would make a wider range of tests possible.

**Comment by aarchiba** [comment]
I think convenience functions are a good idea, but I like the idea of having three separate parts so people can mix and match: setting up TOAs to represent a set of observations, adjusting them so that they give zero residuals, and adding noise.

**Comment by dlakaplan** [comment]
Yes, that's how I have it now:
* `make_fake_toas()`: accepts TOA object
* `make_fake_toas_uniform()`: old behavior (uniformly spaced)
* `make_fake_toas_timfile()`: probably could be replaced with just the first one.

Turns out handling `EFAC`/`EQUAD` is easy with `model.scaled_toa_uncertainty()`.

**Comment by dlakaplan** [comment]
Addresses #1122

**Comment by dlakaplan** [comment]
@paulray : would appreciate your opinion on this (esp. the interface changes).  There is now a CI test for `zima` as well.

**Comment by aarchiba** [comment]
> Yes, that's how I have it now:
> 
>     * `make_fake_toas()`: accepts TOA object
> 
>     * `make_fake_toas_uniform()`: old behavior (uniformly spaced)
> 
>     * `make_fake_toas_timfile()`: probably could be replaced with just the first one.
> 
> 
> Turns out handling `EFAC`/`EQUAD` is easy with `model.scaled_toa_uncertainty()`.

That's not quite what I meant - the three parts I had in mind were:

1. Tools to generate TOAs objects describing various observation schemes (uniform single-frequency, narrowband-like, I'm not sure what)
2. A tool to tweak TOAs to force residuals to zero
3. Tools to add noise to a TOAs object (simple claimed uncertainty, or taking into account the noise model)

The convenience functions would just string a combination of these together to save typing.

**Comment by dlakaplan** [comment]
I'm not sure that the functions you describe are needed.  Or maybe not quite in that breakdown.  For instance, number 3 already exists:
```
        tsim.adjust_TOAs(time.TimeDelta(err))
```
number 2 could be its own routine, but it's also easy enough to just keep that in the first function that I described and have everything call that.  As for the first, I agree that this could be useful, but it also seems like it could easily get very specific and might be better to leave most of the complexity to users (with a few examples).  

But I'm happy to discuss further.

**Comment by aarchiba** [comment]
> I'm not sure that the functions you describe are needed. Or maybe not quite in that breakdown. For instance, number 3 already exists:
> 
> ```
>         tsim.adjust_TOAs(time.TimeDelta(err))
> ```
> 
> number 2 could be its own routine, but it's also easy enough to just keep that in the first function that I described and have everything call that. As for the first, I agree that this could be useful, but it also seems like it could easily get very specific and might be better to leave most of the complexity to users (with a few examples).
> 
> But I'm happy to discuss further.

Well, for example, I've often wanted to generate uniformly-spaced TOAs but don't care what their residuals are. (And that's kind of annoying to do by hand.) More-complicated patterns likewise. I'm not suggesting writing tools to make those as part of this PR, but making room for them in the logical structure seems to make sense, as I can see good reasons for someone to want to use PINT to simulate an observing campaign. 

It sounds like your `make_fake_toas` does nothing but set the residuals to zero, which suggests that maybe it should have a name that says that?

I agree that if you have an instantiation of noise, in seconds, it's pretty easy to add it to TOAs. The third category was meant to be tools for instantiating noise - you could want plain scatter, you could want the EFAC/EQUAD from the model, or you could want full-on ECORR/red noise too.

**Comment by paulray** [comment]
I'm getting an exception and I've tried a few different times but can't make it go away.  I'm running in the tests/datafile directory:
```
% zima --obs @ --freq 0.0 --startMJD 58300.0 --duration 100 --error 100.0 --addnoise --plot --ntoa 100 ngc300nicer.par ngc300fake.tim
INFO: Reading model from ngc300nicer.par [pint.scripts.zima]
WARNING: POSEPOCH not found; using PEPOCH unless set explicitly! [pint.models.astrometry]
INFO: TZRFRQ was 0.0 or None. Setting to infinite frequency. [pint.models.absolute_phase]
INFO: Generating uniformly spaced TOAs [pint.scripts.zima]
INFO: Using CLOCK = TT(TAI), so setting include_bipm = False [pint.simulation]
INFO: Computing TDB columns. [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Applying clock corrections (include_gps = True, include_bipm = False) [pint.toa]
INFO: Special observatory location. No clock corrections applied. [pint.observatory.special_locations]
INFO: Computing TDB columns. [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
INFO: Computing TDB columns. [pint.toa]
INFO: tdb column already exists. Deleting... [pint.toa]
INFO: tdbld column already exists. Deleting... [pint.toa]
INFO: Using EPHEM = DE405 for TDB calculation. [pint.toa]
INFO: Computing PosVels of observatories and Earth, using DE405 [pint.toa]
INFO: Column ssb_obs_pos already exists. Removing... [pint.toa]
INFO: Column ssb_obs_vel already exists. Removing... [pint.toa]
INFO: Column obs_sun_pos already exists. Removing... [pint.toa]
Traceback (most recent call last):
  File "/Users/paulr/env/pint/bin/zima", line 33, in <module>
    sys.exit(load_entry_point('pint-pulsar', 'console_scripts', 'zima')())
  File "/Users/paulr/src/PINT/src/pint/scripts/zima.py", line 85, in main
    ts = pint.simulation.make_fake_toas_uniform(
  File "/Users/paulr/src/PINT/src/pint/simulation.py", line 263, in make_fake_toas_uniform
    return make_fake_toas(ts, model=model, add_noise=add_noise, name=name)
  File "/Users/paulr/src/PINT/src/pint/simulation.py", line 157, in make_fake_toas
    zero_residuals(tsim, model)
  File "/Users/paulr/src/PINT/src/pint/simulation.py", line 67, in zero_residuals
    ts.adjust_TOAs(time.TimeDelta(-r.time_resids))
  File "/Users/paulr/src/PINT/src/pint/toa.py", line 1868, in adjust_TOAs
    self.compute_posvels(self.ephem, self.planets)
  File "/Users/paulr/src/PINT/src/pint/toa.py", line 2279, in compute_posvels
    sun_obs = objPosVel_wrt_SSB("sun", tdb, ephem) - ssb_obs
  File "/Users/paulr/src/PINT/src/pint/solar_system_ephemerides.py", line 167, in objPosVel_wrt_SSB
    pos, vel = astropy.coordinates.get_body_barycentric_posvel(objname, t)
  File "/Users/paulr/env/pint/lib/python3.8/site-packages/astropy/coordinates/solar_system.py", line 337, in get_body_barycentric_posvel
    return _get_body_barycentric_posvel(body, time, ephemeris)
  File "/Users/paulr/env/pint/lib/python3.8/site-packages/astropy/coordinates/solar_system.py", line 291, in _get_body_barycentric_posvel
    for body_p_or_v, p_or_v in zip(body_posvel_bary,
  File "/Users/paulr/env/pint/lib/python3.8/site-packages/jplephem/spk.py", line 232, in generate
    raise OutOfRangeError(
jplephem.exceptions.OutOfRangeError: segment only covers dates 2305424.5 through 2525008.5
```

**Comment by dlakaplan** [comment]
That command works for me, although I don't have that par file.    But:
`zima --obs @ --freq 0.0 --startMJD 58300.0 --duration 100 --error 100.0 --addnoise --plot --ntoa 100 src/pint/data/examples/NGC6440E.par zima.tim`
works.  Can you share your par file?

**Comment by scottransom** [comment]
Is that because of DE405 not covering the dates in question somehow?  Maybe try a different ephem?

**Comment by dlakaplan** [comment]
But the date limits that it's complaining about are like 1599 and 2201.  So something is going wrong somewhere.

**Comment by dlakaplan** [comment]
I can change to DE405 in the NGC6440E par file and it works fine.  Maybe some unset parameter in @paulray 's file is ending up with MJD of 0?

**Comment by paulray** [comment]
The file I used is in tests/datadir in PINT

**Comment by dlakaplan** [comment]
OK, I see what's happening (but don't understand why yet).  The iteration that is supposed to end up at 0 residuals is growing larger rather than smaller, so it's pulling the times to before the SSB calculations are valid.

**Comment by dlakaplan** [comment]
It's to do with the really large F1.

**Comment by paulray** [comment]
Ah, well its a real pulsar so we should be able to handle that case.

**Comment by dlakaplan** [comment]
So as I keep the par file the same and just change the F1 over a range, the max residual behavior changes. Typically it gets smaller with each iteration and converges quickly. For larger |F1|  is takes more iterations. But then larger values still blow up.  
fake_runaway

I'll get to this more on Monday.

**Comment by dlakaplan** [comment]
OK, I figured out part of it.  `residuals.time_resids` uses `get_PSR_freq()`, which doesn't correctly account for spin down.  Replacing that by a calculating involving `model.d_phase_d_toa(toas)` works (mostly) for any F1.  However:
* the apparent frequency calculated this way is a factor of 2 lower than I'd think (but I haven't thought about it hard)
* the iteration now converges, but it doesn't look particularly pretty. The residuals are still too high.

**Comment by paulray** [comment]
Note that `get_PSR_freq()` has a mode to do the calculation correctly:
```
    def get_PSR_freq(self, modelF0=True):
        """Return pulsar rotational frequency in Hz.

        Parameters
        ----------
        modelF0 : bool, optional
            If True, simply read the ``F0`` parameter from the model. If False,
            query the model for the spin period at the time of each TOA.

        Returns
        -------
        freq : Quantity
            Either the single ``F0`` in the model or the spin frequency at the moment of each TOA.
```

**Comment by paulray** [comment]
Also, note that we discussed adding another option to get_PSR_freq() that is in between just grabbing F0 and doing the full d_phase_d_TOA calculation, where it would compute the frequency from the F0 through Fn Taylor series. That would be perfect for this case.

**Comment by dlakaplan** [comment]
Yes, I see.  So maybe it's the fact that this seems to be off by a factor of 2?  Again, will investigate more on Monday

**Comment by dlakaplan** [comment]
old way with a single frequency (from the F0 of the model):
```
In [3]: r.get_PSR_freq()
Out[3]: <Quantity 0.050375 Hz>
```

This is what's needed for a high-F1 pulsar.  But note it's off by a factor of 2:
```
In [4]: r.get_PSR_freq(modelF0=False)
...
Out[4]: 
<Quantity [0.0251875, 0.0251875, 0.0251875, 0.0251875, 0.0251875,
           0.0251875, 0.0251875, 0.0251875, 0.0251875, 0.0251875] Hz>
```
This may also be why `zima` wasn't working on its own (since it used similar routines)?

**Comment by dlakaplan** [comment]
OK, I think I implemented the Taylor-series spin-down as suggested, making use of other pieces.  I don't think the numerical version (which was the alternate to `modelF0`) is right: it's off by a factor of 2 and very noisy.

Now there is:
```
get_PSR_freq(self, calctype="modelF0"):
        """Return pulsar rotational frequency in Hz.

        Parameters
        ----------
        calctype : str, optional
            Type of calculation.  If `calctype` == "modelF0", then simply the ``F0``
            parameter from the model.
            If `calctype` == "numerical", then try a numerical derivative
            If `calctype` == "taylor", evaluate the frequency with a Taylor series
```

and similarly 
```
    def calc_time_resids(self, calctype="modelF0"):
        """Compute timing model residuals in time (seconds).

        Converts from phase residuals to time residuals using several possible ways
        to calculate the frequency.

        Parameters
        ----------
        calctype : str, optional
            Type of calculation.  If `calctype` == "modelF0", then simply the ``F0``
            parameter from the model.
            If `calctype` == "numerical", then try a numerical derivative
            If `calctype` == "taylor", evaluate the frequency with a Taylor series
```
(arguably we might make the Taylor version the default).  

But this works with the high-F1 pulsar Paul had above.  

Screen Shot 2021-08-13 at 8 03 16 PM

**Comment by dlakaplan** [comment]
A few related comments: 
* currently `residuals.get_PSR_freq()` is used by `residuals.get_time_resids()`.  It seems to me that the `get_PSR_freq` command might work better in the `models` package (although it does depend on both the TOAs and the model). 
* In either case we might want to have a persistent `@property` associated with it in case the calculations get lengthy.  But then we'd probably want to store the calculation type.
* We may want to store the calculation type anyway, since you can access the time residuals through the `resids` property and the `time_resids` attribute, but then you don't know what type is returned.
* Is 3 ways 1 (or 2) too many to access the same result?  Function, attribute, and parameter?
* Should we update the `residuals.update()` method to allow it to select the type of residual calculation?  
* Should we change the default frequency calculation to `taylor`?  It seems that this is still pretty fast and there are minimal downsides, except that it can return multiple values (1 per TOA) which may be unexpected.

**Comment by dlakaplan** [comment]
A further note: it looks like the `WidebandDMResiduals` have the `resids` and `resids_value` properties, but they mean different things.

**Comment by dlakaplan** [comment]
Maybe we should use cached properties?

**Comment by dlakaplan** [comment]
Another thought: rather than define functions to convert between phase and frequency, should we use astropy equivalencies?  The following (inserted into `timing_model.py`) allows conversion between s, Hz, and cycles (dimensionless) like `(0.5*u.s).to(u.dimensionless_unscaled, equivalencies=m.equivalencies())`:

(we could also define separate equivalencies for spin and orbital conversions)
```
    def frequency(self, toas=None):
        """Return pulsar rotational frequency in Hz.

        Parameters
        ----------
        toas : pint.toa.TOAs, optional
            TOAs at which frequency should be computed.  If not supplied, uses ``F0`` or ``P0``

        Returns
        -------
        freq : astropy.units.Quantity
            Either the single ``F0`` in the model or the spin frequency at the moment of each TOA.
        """
        if toas is None:
            if "Spindown" in self.components:
                F0 = self.F0.quantity
            elif "P0" in self.params:
                F0 = 1.0 / self.P0.quantity
            else:
                raise AttributeError(
                    "No pulsar spin parameter(e.g., 'F0'," " 'P0') found."
                )
            return F0.to(u.Hz)
        # see Spindown.spindown_phase
        dt = self.get_dt(toas, 0)
        fterms = [0.0 * u.dimensionless_unscaled] + self.get_spin_terms()
        return taylor_horner_deriv(dt, fterms, deriv_order=1).to(u.Hz)

    def equivalencies(self, toas=None):
        freq = self.frequency(toas=toas).to_value(u.Hz)
        return u.equivalencies.Equivalency(
            [
                (u.s, u.Hz, lambda x: 1 / x, lambda x: 1 / x),
                (u.s, u.dimensionless_unscaled, lambda x: x * freq, lambda x: x / freq),
                (
                    u.Hz,
                    u.dimensionless_unscaled,
                    lambda x: x / freq,
                    lambda x: x * freq,
                ),
            ],
            "pulsar",
            {"toas": toas},
        )
```

**Comment by scottransom** [comment]
OMG...  the cycles / freqs dimensionless equivalencies are terribly confusing.  We talked about that when they came out a few years ago and rejected the idea.  There is a bunch of very strange behavior and requirements.

**Comment by dlakaplan** [comment]
@paulray : the test in `tests/test_fake_toas.py::test_zima()` does in fact do what you were asking about: it writes a tim file, reads it back in, and checks the rms.  

It is working fine for me.  I can read in the tim file and compute rms/plot outside of `zima` and it is consistent. Could you be more specific with the test that you did?

**Comment by paulray** [comment]
Here is what I did:
```
cd tests/datafile
zima ngc300nicer.par ngc300fake.tim --addnoise --ntoa 100 --duration 100 --freq 0.0 --startMJD 58000.0 --obs @ --plot
```
That plot looks great.  But, when I do either of the following:
```
tempo2 -gr plk -f ngc300nicer.par ngc300fake.tim
pintk ngc300nicer.par ngc300fake.tim
```
I get a plot like the attached (first plot is from zima, second is from pintk).  Looking closer, I think the error bars are actually fine, but the residuals are coming out about 3x what they were in the plot zima made.
zima
pintk

**Comment by dlakaplan** [comment]
That might be the same error in `pintk` when computing the residuals

**Comment by paulray** [comment]
I just tried the same exercise with the NGC6440E.par and it works fine, so indeed it must be an issue with that particular par file. And, Tempo2 gets the same issue since it also uses the par file F0 for scaling, presumably.

**Comment by paulray** [comment]
Indeed, changing the default calctype to "taylor" in `get_PSR_freq()` fixes the issue.
I think that should be the default, since it will be correct for many more cases than "modelF0", but it does break from the TEMPO/Tempo2 convention.  In general, I think our philosophy has been to make PINT correct where possible, but maintain backwards compatibility for those who need it.

**Comment by dlakaplan** [comment]
Thanks, Paul.  That was my intent as well (default=`taylor`), but I didn't know if that would break things that expected only a single value.

my hope is that I will do a separate PR to fix some of the frequency calculation API

**Comment by dlakaplan** [comment]
I tested the DM fitting we discussed.  It works fine (I believe).  What I did was:
* look at the NGC6440E dataset.  The last 8 TOAs are different freqs on the same day, so I selected only those
* I fit the model with only DM free to those data
* I then generated N (=30) fake TOA sets matching the 8 multi-free TOAs.  For each I again fit for only DM
* I verified that the rms of the fitted DM values is comparable to the uncertainty on the DM value from the original (real) fit.

I added a test for this (`test_fake_toas::test_fake_DMfit()`)

**Comment by paulray** [comment]
Any reason not to merge this and #1129 now?

**Comment by dlakaplan** [comment]
The latest push broke a notebook, so I'm fixing that now

**Comment by paulray** [comment]
Cool, can you also add a CHANGELOG entry?

**Comment by dlakaplan** [comment]
when we merge I will

**Comment by dlakaplan** [comment]
I think this is ready to merge (@paulray or @scottransom )

File: src/pint/residuals.py [hunk 1]
```diff
@@ -267,7 +270,23 @@ def get_PSR_freq(self, modelF0=True):
...
+            dt = self.model.get_dt(self.toas, 0)
+            # if the model is defined through F0, F1, ...
+            if "F0" in self.model.params:
+                fterms = [0.0 * u.dimensionless_unscaled] + self.model.get_spin_terms()
+
+            # otherwise assume P0, PDOT
+            else:
+                F0 = 1.0 / self.model.P0.quantity
+                if "PDOT" in self.model.params:
+                    F1 = -self.model.PDOT.quantity / self.model.P0.quantity ** 2
```
Comment (paulray) [inline src/pint/residuals.py:284]: I know this computation is sort of obvious, but it might be good form to use `derived_quantities.p_to_f()`

**Review by paulray — APPROVED** [review]

**Pull request merged**